web/src/stream_data.ts

```diff
@@ -619,7 +623,14 @@ export function can_access_topic_history(sub: StreamSubscription): boolean {
 }
 
 export function can_preview(sub: StreamSubscription): boolean {
-    return sub.subscribed || !sub.invite_only || sub.previously_subscribed;
+    if (sub.previously_subscribed) {
+        return true;
+    }
+    const user_has_content_access = has_content_access(sub);
+    if (user_has_content_access && sub.invite_only && !sub.history_public_to_subscribers) {
```

Comment:
I think this should probably just be checking `!sub.history_public_to_subscribers`; it's not relevant to this logic whether the channel is invite-only. (I know we don't in general allow protected history without private channels, but if that changes, this logic would need an extra update. And I think in any case, it's clearest to be precise about what's relevant).

Oh, this can then be cleaned to just `if(!sub.history_public_to_subscribers) { return false)` -- if `!user_has_content_access`, it'd return false next anyway.